controllers/clickhouseuser_controller.go

```diff
@@ -160,32 +265,108 @@ func (h *clickhouseUserHandler) get(ctx context.Context, avnGen avngen.Client, o
 	}
 
-	secret := newSecret(user, stringData, false)
+	goalSecret := newSecret(user, stringData, false)
 
 	meta.SetStatusCondition(&user.Status.Conditions,
 		getRunningCondition(metav1.ConditionTrue, "CheckRunning",
 			"Instance is running on Aiven side"))
-
 	metav1.SetMetaDataAnnotation(&user.ObjectMeta, instanceIsRunningAnnotation, "true")
-	return secret, nil
+
+	if user.NoSecret() {
+		r.Recorder.Event(user, corev1.EventTypeNormal, eventConnInfoSecretCreationDisabled, "connInfoSecretTargetDisabled is true, secret will not be created")
+		return nil
+	}
+
+	// `CreateOrUpdate` will populate this secret by calling Get
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      goalSecret.Name,
+			Namespace: goalSecret.Namespace,
+		},
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, secret, func() error {
+		if secret.Data == nil {
+			secret.Data = make(map[string][]byte)
+		}
+
+		// copy data from our goalSecret's StringData.
+		// this handles both Create and Update
+		if goalSecret.StringData != nil {
+			secret.Data = make(map[string][]byte) // clear existing data
+			for key, value := range goalSecret.StringData {
+				secret.Data[key] = []byte(value)
+			}
+		}
+
+		secret.Labels = goalSecret.Labels
+		secret.Annotations = goalSecret.Annotations
+
+		return controllerutil.SetControllerReference(user, secret, r.Client.Scheme())
+	})
+
+	return err
 }
 
-func (h *clickhouseUserHandler) checkPreconditions(ctx context.Context, avnGen avngen.Client, obj client.Object) (bool, error) {
-	user, err := h.convert(obj)
+// Delete implements AivenClient for ClickhouseUser.
+// It mirrors the current delete behaviour used in reconcile2.
+func (r *ClickhouseUserAivenClient) Delete(ctx context.Context, user *v1alpha1.ClickhouseUser) error {
+	avnGen, err := r.newAivenClient(ctx, user)
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	meta.SetStatusCondition(&user.Status.Conditions,
-		getInitializedCondition("Preconditions", "Checking preconditions"))
+	// Not processed yet
+	if user.Status.UUID == "" {
+		return nil
+	}
+
+	// skip deletion for built-in users that cannot be deleted
+	if isBuiltInUser(user.Name) {
+		// built-in users like 'default' cannot be deleted, this is expected behavior
+		// we consider this a successful deletion since we can't and shouldn't delete built-in users
+		return nil
+	}
 
-	return checkServiceIsOperational(ctx, avnGen, user.Spec.Project, user.Spec.ServiceName)
+	err = avnGen.ServiceClickHouseUserDelete(ctx, user.Spec.Project, user.Spec.ServiceName, user.Status.UUID)
+	if !isNotFound(err) {
+		return err
+	}
+
+	return nil
 }
 
-func (h *clickhouseUserHandler) convert(i client.Object) (*v1alpha1.ClickhouseUser, error) {
-	user, ok := i.(*v1alpha1.ClickhouseUser)
-	if !ok {
-		return nil, fmt.Errorf("cannot convert object to ClickhouseUser")
+func newAivenClient(c Controller) func(ctx context.Context, obj v1alpha1.AivenManagedObject) (avngen.Client, error) {
```

Comment:
The same thing with a possibly extra polymorphism: https://github.com/aiven/aiven-operator/pull/1067/files#r2533483150
On top of that, `newAivenClient` closes over the whole `Controller`, while it really only needs a small subset of its fields. It might be cleaner to copy only the minimal data.